vms/avm/states/state.go

```diff
@@ -848,3 +866,65 @@ func skipTo(iter database.Iterator, targetKey []byte) error {
 		}
 	}
 }
+
+func (s *state) Checksums() (ids.ID, ids.ID) {
+	return s.txChecksum, s.utxoState.Checksum()
+}
+
+func (s *state) initTxChecksum() error {
+	if !s.trackChecksum {
+		return nil
+	}
+
+	txIt := s.txDB.NewIterator()
+	defer txIt.Release()
+	statusIt := s.statusDB.NewIterator()
+	defer statusIt.Release()
+
+	statusHasNext := statusIt.Next()
+	for txIt.Next() {
+		txIDBytes := txIt.Key()
+		if statusHasNext { // if status was exhausted, everything is accepted
+			statusIDBytes := statusIt.Key()
+			if bytes.Equal(txIDBytes, statusIDBytes) { // if the status key doesn't match this was marked as accepted
+				statusInt, err := database.ParseUInt32(statusIt.Value())
+				if err != nil {
+					return err
+				}
+
+				statusHasNext = statusIt.Next() // we processed the txID, so move on to the next status
+
+				if choices.Status(statusInt) != choices.Accepted { // the status isn't accepted, so we skip the txID
+					continue
+				}
+			}
+		}
+
+		txID, err := ids.ToID(txIDBytes)
+		if err != nil {
+			return err
+		}
+
+		s.updateTxChecksum(txID)
+	}
+
+	if statusHasNext {
+		return errStatusWithoutTx
+	}
+
+	errs := wrappers.Errs{}
+	errs.Add(
+		txIt.Error(),
+		statusIt.Error(),
+	)
+	return errs.Err
+}
+
+func (s *state) updateTxChecksum(modifiedID ids.ID) {
+	if !s.trackChecksum {
+		return
+	}
+	for i, b := range modifiedID {
+		s.txChecksum[i] ^= b
+	}
```

Comment:
This is duplicated in the UTXO state as well... We could add an `XOR` to the `ids.ID` type... I wasn't sure if the signature should be:

- `(*ids.ID) XOR(ids.ID)` <- modifies the first ID
- `(ids.ID) XOR(ids.ID) ids.ID` <- returns the result of XORing the IDs

So I ended up just not making the helper... But if someone wants to decide what to do I will follow it.

Reply:
> (ids.ID) XOR(ids.ID) ids.ID <- returns the result of XORing the IDs

fwiw I like this better